testgrid/cmd/configurator/main.go

```diff
@@ -144,18 +160,67 @@ func announceProwChanges(ctx context.Context, pca *prowConfig.Agent, channel cha
 
 func readConfig(paths []string) (*Config, error) {
 	var c Config
-	for _, file := range paths {
-		b, err := ioutil.ReadFile(file)
+	err := walkForYAMLFiles(paths, func(path string, info os.FileInfo) error {
+		// Read YAML file and update config
+		b, err := ioutil.ReadFile(path)
 		if err != nil {
-			return nil, fmt.Errorf("failed to read %s: %v", file, err)
+			return fmt.Errorf("failed to read %s: %v", path, err)
 		}
 		if err = c.Update(b); err != nil {
-			return nil, fmt.Errorf("failed to merge %s into config: %v", file, err)
+			return fmt.Errorf("failed to merge %s into config: %v", path, err)
 		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
 	}
 	return &c, nil
 }
 
+// walks through paths and directories, calling the passed function on each YAML file
+// future modifications to what Configurator sees as a "config file" can be made here
+func walkForYAMLFiles(paths []string, callFunc func(path string, info os.FileInfo) error) error {
+	for _, path := range paths {
+		_, err := os.Stat(path)
+		if err != nil {
+			return fmt.Errorf("Failed status call on %s: %v", path, err)
+		}
+
+		err = filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				logrus.WithError(err).Errorf("walking path %q.", path)
+				// bad file should not stop us from parsing the directory
+				return nil
+			}
+
+			if strings.HasPrefix(info.Name(), "..") {
```

Comment:
I'm curious what files are being added with a `..` prefix?

Reply:
I took a number of hints from the way that Prow does this to determine the best way to traverse and monitor these files

When Prow is digging through jobs, it ignores ".." directories in this way. [Example](https://github.com/kubernetes/test-infra/blob/a3c842850be5c9c8681d38ce52c03163e2d918ca/prow/config/config.go#L504-L506)
Honestly, I can't tell you exactly why; filepath.Walk ignores the obvious ".." link to go up a directory. From the comment, it seems that kubernetes volumes contain actual files prefixed by ".." that need to be explicitly ignored

@krzyzacy , do you know why this code is there in Prow? [Commit](https://github.com/kubernetes/test-infra/commit/41f7bc25f51bce38b3cc89f3bfac2544789c0c34)

Reply:
ref https://github.com/kubernetes/test-infra/pull/8454#issuecomment-400162513

It's a Kubernetes storage specific thing (it auto creates subdir with `..****` in a configmap), and probably has nothing to do with testgrid here.